From e36b616bcbc19eded80680f6b7057b6af3b3e226 Mon Sep 17 00:00:00 2001
From: Bubun Das <bubundas17@gmail.com>
Date: Sat, 8 Aug 2026 13:06:35 +0000
Subject: [PATCH 17/20] SystemUI: Show UDFPS dim layer during AOD fingerprint
 press

While the display is dozing the dim layer is hidden because the shade is
not touchable, so the global HBM used to illuminate the optical sensor
flashes the whole screen up unmasked. Show the dim layer for the
duration of the fingerprint press so the AOD press stays dark and only
the sensor circle lights up, matching lockscreen behavior.
---
 .../systemui/biometrics/UdfpsController.java  |  8 ++++-
 .../biometrics/UdfpsControllerOverlay.kt      |  4 +++
 .../systemui/biometrics/UdfpsHelper.kt        | 29 +++++++++++++++----
 3 files changed, 34 insertions(+), 7 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
index 5bc6bef75cfd..159401a769c1 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
@@ -660,7 +660,10 @@ public class UdfpsController implements DozeReceiver, Dumpable {
             // It's possible on some devices to get duplicate touches from both doze and the
             // normal touch listener. Don't reset the down in this case to avoid duplicate downs
             if (!mIsAodInterruptActive) {
-                mOnFingerDown = false;
+        mOnFingerDown = false;
+        if (mOverlay != null) {
+            mOverlay.setFingerDown(false);
+        }
             }
         }
 
@@ -1268,6 +1271,9 @@ public class UdfpsController implements DozeReceiver, Dumpable {
             mDeviceEntryFaceAuthInteractor.onUdfpsSensorTouched();
         }
         mOnFingerDown = true;
+        if (mOverlay != null) {
+            mOverlay.setFingerDown(true);
+        }
         mFingerprintManager.onPointerDown(requestId, mSensorProps.sensorId, pointerId, x, y,
                 minor, major, orientation, time, gestureStart, isAod);
 
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsControllerOverlay.kt b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsControllerOverlay.kt
index ac2b91ce0380..05c2c1672595 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsControllerOverlay.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsControllerOverlay.kt
@@ -136,6 +136,10 @@ constructor(
         return overlayTouchView
     }
 
+    fun setFingerDown(fingerDown: Boolean) {
+        udfpsHelper?.setFingerDown(fingerDown)
+    }
+
     private var overlayParams: UdfpsOverlayParams = UdfpsOverlayParams()
     private var sensorBounds: Rect = Rect()
 
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsHelper.kt b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsHelper.kt
index 66b8ce972eaa..5a91dccbe361 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsHelper.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsHelper.kt
@@ -218,15 +218,32 @@ class UdfpsHelper(
         }
     }
 
+    private var isShadeTouchable: Boolean = false
+
+    private var isFingerDown: Boolean = false
+
+    fun setFingerDown(fingerDown: Boolean) {
+        if (!isKeyguard || isFingerDown == fingerDown) {
+            return
+        }
+        isFingerDown = fingerDown
+        updateDimLayerVisibility()
+    }
+
+    private fun updateDimLayerVisibility() {
+        view.isVisible = isShadeTouchable || isFingerDown
+        if (view.isVisible) {
+            brightnessToAlpha()
+            alphaAnimator.cancel()
+            alphaAnimator.start()
+        }
+    }
+
     private suspend fun listenForShadeTouchability(scope: CoroutineScope): Job {
         return scope.launch {
             shadeInteractor.isShadeTouchable.collect {
-                view.isVisible = it
-                if (view.isVisible) {
-                    brightnessToAlpha()
-                    alphaAnimator.cancel()
-                    alphaAnimator.start()
-                }
+                isShadeTouchable = it
+                updateDimLayerVisibility()
             }
         }
     }
-- 
2.43.0

